packages/db/src/project/loadCompile/sources.ts

```diff
@@ -5,43 +5,194 @@
+          "generatedSources",
+          "deployedGeneratedSources"
+        ] as const) {
+          const generatedSources = contract[property];
+          if (generatedSources) {
+            for (const [
+              generatedSourceIndex,
+              source
+            ] of generatedSources.entries()) {
+              yield {
+                input: source,
+                breadcrumb: {
+                  compilationIndex,
+                  sourcePointer: {
+                    contractIndex,
+                    property,
+                    generatedSourceIndex
+                  }
+                }
+              };
+            }
+          }
+        }
+      }
+    }
+  },
+
+  find({ inputs, breadcrumb }) {
+    const { compilationIndex, sourcePointer } = breadcrumb;
+    const compilation = inputs[compilationIndex];
+
+    if ("generatedSourceIndex" in sourcePointer) {
+      const { contractIndex, property, generatedSourceIndex } = sourcePointer;
+
+      // @ts-ignore
+      return compilation.contracts[contractIndex][property][
+        generatedSourceIndex
+      ];
+    }
+
+    const { sourceIndex } = sourcePointer;
+    return compilation.sources[sourceIndex];
+  },
+
+  extract({ input }) {
+    return {
+      contents: input.contents,
+      sourcePath: "sourcePath" in input ? input["sourcePath"] : input["name"]
+    };
   },
 
   *process({ entries }) {
     return yield* resources.load("sources", entries);
   },
 
-  convert<_I, _O>({ result, input: source }) {
+  initialize<_I, _O>({ inputs }) {
+    return inputs.map(compilation => ({
+      ...compilation
+    }));
+  },
+
+  convert<_I, _O>({ result, input }) {
     return {
-      ...source,
+      ...input,
       db: {
-        ...source.db,
         source: result
       }
     };
+  },
+
+  merge<I, O>({ outputs, breadcrumb, output }) {
+    const { compilationIndex, sourcePointer } = breadcrumb;
+
+    const compilationsBefore = outputs.slice(0, compilationIndex);
+    const compilation = outputs[compilationIndex];
+    const compilationsAfter = outputs.slice(compilationIndex + 1);
+
+    if ("generatedSourceIndex" in sourcePointer) {
+      const { contractIndex, property, generatedSourceIndex } = sourcePointer;
+      const contractsBefore = compilation.contracts.slice(0, contractIndex);
+      const contract = compilation.contracts[contractIndex];
+      const contractsAfter = compilation.contracts.slice(contractIndex + 1);
+
+      const generatedSourcesBefore = contract[property].slice(
+        0,
+        generatedSourceIndex
+      );
+      const generatedSource = output;
+      const generatedSourcesAfter = contract[property].slice(
+        generatedSourceIndex + 1
+      );
+
+      return [
+        ...compilationsBefore,
+        {
+          ...compilation,
+          contracts: [
+            ...contractsBefore,
+            {
+              ...contract,
+              [property]: [
+                ...generatedSourcesBefore,
```

Comment:
this merge is getting a little deep! seems a reasonable way to do this though.

Reply:
Yeah :/